### ISSUE TITLE
Not quite get updateBinaryGradWeight()

### ISSUE THREAD

**Issue by jiang12zhiming** [jiecaoyu/XNOR-Net-PyTorch#28]
Thanks so much for sharing, I am new to deep learning, and I learned a lot from your code. 
I have a question, I kind get 
m = weight.norm(1, 3, keepdim=True).sum(2, keepdim=True).sum(1, keepdim=True).div(n).expand(s)
is to calculate 
<img width="278" alt="screen shot 2018-06-28 at 1 06 26 pm" src="https://user-images.githubusercontent.com/12436972/42049471-176ef2c6-7ad4-11e8-8dd9-8bced733bc49.png">
Next step why you use  m * self.target_modules[index].grad.data 

Finally how do you get new weight grad data.  
self.target_modules[index].grad.data = m.add(m_add).mul(1.0-1.0/s[1]).mul(n)

Could you explain more?  for example which part is the 
<img width="86" alt="screen shot 2018-06-28 at 1 10 42 pm" src="https://user-images.githubusercontent.com/12436972/42049700-ba9e806a-7ad4-11e8-84be-cc6cc8d8944a.png">
<img width="55" alt="screen shot 2018-06-28 at 1 10 49 pm" src="https://user-images.githubusercontent.com/12436972/42049705-be0f53f0-7ad4-11e8-963d-1d700b808b76.png">
<img width="30" alt="screen shot 2018-06-28 at 1 10 55 pm" src="https://user-images.githubusercontent.com/12436972/42049706-c047069a-7ad4-11e8-84b2-79601ebfa459.png">

**Comment by jiecaoyu** [comment]
Hi, @jiang12zhiming 

1.  [line 78 and line 79](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/ImageNet/networks/util.py#L78) perform this

2. It is the ```self.target_modules[index].grad.data``` [here](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/ImageNet/networks/util.py#L80)

3. It is the ```m``` calculated by [line 73 to line 77](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/ImageNet/networks/util.py#L73)

**Comment by jiang12zhiming** [comment]
Thank you so much for the quick reply, I have just one more question. 
in line self.target_modules[index].grad.data = m.add(m_add).mul(1.0-1.0/s[1]).mul(n)

you multiple a scale .mul(1.0-1.0/s[1]).mul(n)
which is 
(1 - 1 / input channel of weight matrix) * output channel weight matrix. 

<img width="537" alt="screen shot 2018-06-28 at 1 50 16 pm" src="https://user-images.githubusercontent.com/12436972/42051488-35627e28-7ada-11e8-8618-72d224b9f4dd.png">

My question is
I am not sure why you have this scale, which part of the formula this scale refer?

**Comment by jiecaoyu** [comment]
```.mul(1.0-1.0/s[1])``` is the gradient for the ```meancenterConvParams()```. 

```.mul(n)``` is used to mitigate the effect of weight decay.

**Comment by jiang12zhiming** [comment]
Thank you so much.
